core/src/impl/Kokkos_Stacktrace.cpp

```diff
@@ -0,0 +1,248 @@
+  size_t start             = found_parenthesis + 1;
+  if (found_parenthesis == std::string::npos) start = 0;
+
+  std::string s = name.substr(start, found_end - start);
+
+  if (s.length() != 0) {
+    int status          = 0;
+    char* output_buffer = nullptr;
+    size_t length       = s.length();
+    char* d = abi::__cxa_demangle(s.c_str(), output_buffer, &length, &status);
+    if (d != nullptr) {
+      s = d;
+      free(d);
+    }
+  }
+
+  // Special cases for "main" and "start" on Mac
+  if (s.length() == 0) {
+    if (name == "main" || name == "start") {
+      s = name;
+    }
+  }
+  return s;
+#endif  // KOKKOS_ENABLE_CXXABI
+}
+
+class Stacktrace {
+ public:
+  Stacktrace()                  = delete;
+  Stacktrace(const Stacktrace&) = delete;
+  Stacktrace& operator=(const Stacktrace&) = delete;
+  Stacktrace(Stacktrace&&)                 = delete;
+  Stacktrace& operator=(Stacktrace&&) = delete;
+  ~Stacktrace()                       = delete;
+
+  // These are public only to avoid wasting an extra stacktrace line.
+  // See save_stacktrace below.
+  static constexpr int capacity = 100;
+  static void* buffer[capacity];
+  static int length;
+
+  static std::vector<std::string> lines() {
+    char** symbols = backtrace_symbols(buffer, length);
+    if (symbols == nullptr) {
+      return {};
+    } else {
+      std::vector<std::string> trace(length);
+      for (int i = 0; i < length; ++i) {
+        if (symbols[i] != nullptr) {
+          trace[i] = std::string(symbols[i]);
+        }
+      }
+      free(symbols);
+      return trace;
+    }
+  }
+};
+
+int Stacktrace::length = 0;
+void* Stacktrace::buffer[Stacktrace::capacity];
+
+void save_stacktrace() {
+  Stacktrace::length = backtrace(Stacktrace::buffer, Stacktrace::capacity);
+}
+
+size_t find_first_non_whitespace(const std::string& s, const size_t start_pos) {
+  constexpr size_t num_ws_chars = 3;
+  const char ws_chars[]         = "\n\t ";
+  return s.find_first_not_of(ws_chars, start_pos, num_ws_chars);
+}
+
+size_t find_first_whitespace(const std::string& s, const size_t start_pos) {
+  constexpr size_t num_ws_chars = 3;
+  const char ws_chars[]         = "\n\t ";
+  return s.find_first_of(ws_chars, start_pos, num_ws_chars);
+}
+
+template <class Callback>
+void for_each_token(const std::string& s, Callback c) {
+  size_t cur = find_first_non_whitespace(s, 0);
+  while (cur != std::string::npos) {
+    const size_t end   = find_first_whitespace(s, cur);
+    const bool last    = (end == std::string::npos);
+    const size_t count = last ? end : size_t(end - cur);
+    c(s.substr(cur, count), last);
+    cur = find_first_non_whitespace(s, end);
+  }
+}
+
+// Search the whole backtrace, column by column, for "main".
+// This tells us what column has the function names.
+// While we're doing that, figure out the longest column,
+// so we can compute spacing correctly.
+
+std::tuple<bool, size_t, std::vector<size_t>> find_main_column(
```

Comment:
Don't use tuple for something that you have to document the meaning of. Use a simple struct with names; something like:

```c++
struct find_main_column_result {
  bool found_main;
  size_t main_col;
  std::vector<size_t> main_col_lens;
};
```

Reply:
I can't believe it. David proposing something which reduces the number of C++ism we need (this also gets rid of the get thing).

Reply:
Done.